### PR TITLE
add Jenkinsfile

### PR DESCRIPTION
**Pull request by serhanwbahar**

### CODE REVIEW

**Comment by serhanwbahar** [comment]
Waiting to discuss with @amirhouieh. He is out-of-office until Monday.

**Comment by serhanwbahar** [comment]
We discussed this with @amirhouieh. Related to https://github.com/status-im/infra-misc/issues/108#issuecomment-1446449960

**Comment by serhanwbahar** [comment]
@jakubgs I updated the Jenkinsfile regarding our latest discussion

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        beforeAgent true
+      }
+      steps {
+        git branch: 'main', url: 'https://github.com/acid-info/logos-site-builder'
+      }
+    }
+
+    stage('Clone content repository') {
+      steps {
+        git branch: 'main', url: "https://github.com/acid-info/logos.co"
```
Comment (jakubgs) [inline Jenkinsfile:34]: Why are you doing this? Think about where the `Jenkinsfile` the Jenkins instance is running is coming from.

Have you tested this `Jenkinsfile` in a CI run?

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      }
+    }
+
+    stage('Update environment') {
+      steps {
+        script {
+          def envFile = '.env'
+          def content = "CONTENT_SOURCE_TYPE=git\nCONTENT_SOURCE_URL=\"https://github.com/acid-info/logos.co\""
```
Comment (jakubgs) [inline Jenkinsfile:42]: This content is not dynamic. Why are you creating it dynamically then?
There are no variables that are job-specific that you need to include her. Just crate that file.
Either in `ci/.env` or `.env.ci` and then symlink it at build time. You are over-complicating this.

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          writeFile file: "${envFile}", text: "${content}"
+        }
+      }
+    }
+
+    stage('Build') {
+      steps {
+        sh 'yarn install'
```
Comment (jakubgs) [inline Jenkinsfile:50]: This can be it's own stage.

File: Jenkinsfile (hunk 2 above)
Comment (jakubgs) [reply Jenkinsfile:42]: Also, does this have to be a file? Why not just set those in `environment` section?

**Comment by jakubgs** [comment]
And again, jenkins doesn't want to create the webhook:

image

Annoying.

**Comment by jakubgs** [comment]
I created it manually:

image

File: Jenkinsfile (hunk 1 above)
Comment (serhanwbahar) [reply Jenkinsfile]: Fixed and tested in a CI Run:
Screenshot 2023-03-07 at 12 14 57

File: Jenkinsfile (hunk 2 above)
Comment (serhanwbahar) [reply Jenkinsfile]: fixed

File: Jenkinsfile (hunk 3 above)
Comment (serhanwbahar) [reply Jenkinsfile]: fixed

**Comment by jakubgs** [comment]
You should probably try using `CONTENT_SOURCE_TYPE=local_folder`.
https://github.com/acid-info/logos-site-builder/blob/5d588427171da4562e92417b9c56a107927c8356/scripts/download-source.js#L38-L49

**Comment by serhanwbahar** [comment]
I'll fix the error tomorrow:

```
Error downloading content from source,  https://github.com/acid-info/logos.co/archive/refs/heads/develop.zip
[Error: ENOENT: no such file or directory, lstat 'https://github.com/acid-info/logos.co/archive/refs/heads/develop.zip'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'lstat',
  path: 'https://github.com/acid-info/logos.co/archive/refs/heads/develop.zip'
}
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[Error: ENOENT: no such file or directory, open '/home/jenkins/workspace/website/logos.co/docs/config.json'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/home/jenkins/workspace/website/logos.co/docs/config.json'
}
error Command failed with exit code 1.
```

**Comment by jakubgs** [comment]
Look at the code. Look at what `local_folder` actually means.

**Comment by amirhouieh** [comment]
@serhanwbahar CONTENT_SOURCE_TYPE should be set to git if the url is referring to a remote zip file. I think problem lays somewhere else. Let's catch up tomorrow on a short call to look at it together if you want.

**Comment by amirhouieh** [comment]
- `CONTENT_SOURCE_TYPE` should always be set to `git` if `CONTENT_SOURCE_URL` contains a remote zip url. `CONTENT_SOURCE_TYPE="local_folder"` is only for local development purposes when the source is a folder on your local machine. 
- `CONTENT_SOURCE_URL` should actually be able to get two different values per enviroment or domain; for example one for `logos.co` that refers to the master content repo `...master.zip` and one for dev.logos.co that refers to the development branch of content repo. This is because the site owner should be able to see their changes live via dev.logos.co before going to production domain. This was easy via Vercel because it let you to create env vars with same name but assign them different values based on domain. I am not familiar with Jekinsfile, but this is how I would solve it:

```
//in environment configs, we define two new variables  
CONTENT_SOURCE_URL_DEV=https://github.com/acid-info/logos.co/archive/refs/heads/develop.zip
CONTENT_SOURCE_URL_PROD=https://github.com/acid-info/logos.co/archive/refs/heads/master.zip

 
// and then somewhere in your jekinsfile you overwrite `CONTENT_SOURCE_URL` conditionally based on the environment.
if(dev)
    CONTENT_SOURCE_URL=CONTENT_SOURCE_URL_DEV
else 
    CONTENT_SOURCE_URL=CONTENT_SOURCE_URL_PROD


```

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    /* dev page settings */
+    DEV_SITE = 'dev.logos.co'
+    DEV_HOST = 'jenkins@node-01.do-ams3.sites.misc.statusim.net'
+    SCP_OPTS = 'StrictHostKeyChecking=no'
+  }
+
+  stages {
+    stage('Clone framework repository') {
+      steps {
+        git branch: 'v0', url: 'https://github.com/acid-info/logos-site-builder'
```
Comment (jakubgs) [inline Jenkinsfile:27]: One question to ask is, where is this being checked out into? Seems like it's workspace root, but that's probably wrong.

We'd want to check this out into a subfolder.

**Comment by serhanwbahar** [comment]
@jakubgs I've tried some troubleshooting steps to see if we can do that without changing the source code. I think we cannot do that without changing the source code. I need your thoughts about my steps to see whether I missed anything.

Also, what should be the next step? Trying to change the source code or using it with a zip file?

**Comment by jakubgs** [comment]
Using the ZIP is dumb and ugly. We already have the repo checked out, why would we then fetch the ZIP?

Not to mention, while the build starts the could be another commit, creating a race condition where in some cases the build will use a different commit than the one that started the build. Which is also just bad.

**Comment by jakubgs** [comment]
I made it work. It's not complex:
```
 > ll
total 3.0K
drwxr-xr-x  6 jakubgs jakubgs 17 Mar  9 18:27 logos.co
drwxr-xr-x 18 jakubgs jakubgs 28 Mar  9 18:40 logos-site-builder

 > cd logos-site-builder 
 > export CONTENT_SOURCE_TYPE=local_folder
 > export CONTENT_SOURCE_URL=../logos.co
 > yarn build
yarn run v1.22.19
$ yarn build:data && next build && next export
$ node scripts/index.js
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
info  - Checking validity of types  
info  - Creating an optimized production build .Browserslist: caniuse-lite is outdated. Please run:
```
You can use `checkoutToSubdirectory()` like we do in [`go-waku` Jenkinsfile](https://github.com/status-im/go-waku/blob/349e22bbe8ca9bc7a53712dcebfb4ed168b3f3d6/ci/Jenkinsfile.linux#L16-L17) to checkout the site repo into it's own subfolder. And then in a separate step checkout the builder to another subfolder. Go into that and build using the vars above.

**Comment by serhanwbahar** [comment]
> I made it work. It's not complex:
> 
> ```
>  > ll
> total 3.0K
> drwxr-xr-x  6 jakubgs jakubgs 17 Mar  9 18:27 logos.co
> drwxr-xr-x 18 jakubgs jakubgs 28 Mar  9 18:40 logos-site-builder
> 
>  > cd logos-site-builder 
>  > export CONTENT_SOURCE_TYPE=local_folder
>  > export CONTENT_SOURCE_URL=../logos.co
>  > yarn build
> yarn run v1.22.19
> $ yarn build:data && next build && next export
> $ node scripts/index.js
> Browserslist: caniuse-lite is outdated. Please run:
>   npx browserslist@latest --update-db
>   Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
> info  - Checking validity of types  
> info  - Creating an optimized production build .Browserslist: caniuse-lite is outdated. Please run:
> ```
> 
> You can use `checkoutToSubdirectory()` like we do in [`go-waku` Jenkinsfile](https://github.com/status-im/go-waku/blob/349e22bbe8ca9bc7a53712dcebfb4ed168b3f3d6/ci/Jenkinsfile.linux#L16-L17) to checkout the site repo into it's own subfolder. And then in a separate step checkout the builder to another subfolder. Go into that and build using the vars above.
> 
> It's dumb because we actually allow the silly script they have copy the source locally FOR NO REASON, but at least it works.

Thanks for this! I should have tested it locally (as you said); it would have been much faster; anyway, I'll do it as you described.

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    GIT_COMMITTER_NAME = 'status-im-auto'
+    GIT_COMMITTER_EMAIL = 'auto@status.im'
+    /* dev page settings */
+    DEV_SITE = 'dev.logos.co'
+    DEV_HOST = 'jenkins@node-01.do-ams3.sites.misc.statusim.net'
+    SCP_OPTS = 'StrictHostKeyChecking=no'
+  }
+
+  stages {
+    stage('Clone framework repository') {
```
Comment (jakubgs) [inline Jenkinsfile:26]: Far too long. Just call it `Clone Builder`.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      steps {
+        sshagent(credentials: ['jenkins-ssh']) {
+          sh """
+            rsync -e 'ssh -o ${SCP_OPTS}' -r --delete public/. \
+              ${env.DEV_HOST}:/var/www/${env.DEV_SITE}/
+          """
+        }
+      }
+    }
+  }
```
Comment (jakubgs) [inline Jenkinsfile:64]: We need some cleanup to not pollute future builds with previous builds:s
```suggestion
  }

  post {
    cleanup { cleanWs() }
  }
```

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    /* manage how many builds we keep */
+    checkoutToSubdirectory('src/')
+    buildDiscarder(logRotator(
+      numToKeepStr: '20',
+      daysToKeepStr: '30',
+    ))
+  }
+
+  environment {
+    CONTENT_SOURCE_TYPE  = 'local_folder'
```
Comment (jakubgs) [inline Jenkinsfile:15]: Maybe add a comment explaining a bit why this is here and what it does?

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,65 @@
+pipeline {
+  agent { label 'linux' }
+
+  options {
+    disableConcurrentBuilds()
+    /* manage how many builds we keep */
```
Comment (jakubgs) [inline Jenkinsfile:6]: You moved the comment. Now it doesn't apply to the line it should apply to.

File: Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    checkoutToSubdirectory('src/')
```
Comment (jakubgs) [inline Jenkinsfile:7]: Add a short comment as to why we need to checkout using a subfolder.

**Review by jakubgs — APPROVED** [review]
Almost there. Once you are done please squash it all.

File: Jenkinsfile [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    /* manage how many builds we keep */
+    buildDiscarder(logRotator(
+      numToKeepStr: '20',
+      daysToKeepStr: '30',
+    ))
+  }
+
+  environment {
+    /* checkout the source code, use local folder */
```
Comment (jakubgs) [inline Jenkinsfile:16]: No, that's not a good comment. A good comment explains WHY you are doing something, not WHAT you are doing.
Look at this:
```js
  /* Make request to API */
  api.request()
```
How is this comment useful in any way? It is not. It provides no additional information.

The commend we need is more like:
```suggestion
    /* Mode of logos-site-builder for copying site source from already checked out repo.
     * TODO: Avoid copying anything at all, make checkout site of into `docs` folder work. */
```

File: Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,71 @@
+pipeline {
+  agent { label 'linux' }
+
+  options {
+    disableConcurrentBuilds()
+    /* checkout logos repo to use site builder */
```
Comment (jakubgs) [inline Jenkinsfile:6]: Same here. This is not a useful comment. Do not describe WHAT code does, describe WHY it does it.
```suggestion
    /* Necessary for logos-side-builder local_folder source type. */
```

**Comment by serhanwbahar** [comment]
@jakubgs It seems it works fine now. If this approach is ok, I'll squash it and expand my commit comment to finalise it.

File: Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    /* dev page settings */
+    DEV_SITE = 'dev.logos.co'
+    DEV_HOST = 'jenkins@node-01.do-ams3.sites.misc.statusim.net'
+    SCP_OPTS = 'StrictHostKeyChecking=no'
+  }
+
+  stages {
+    stage('Clone Builder') {
+      steps {
+        dir('builder/') {
```
Comment (jakubgs) [inline Jenkinsfile:31]: You don't need the trailing slashes.

File: Jenkinsfile [hunk 13]
```diff
@@ -0,0 +1,80 @@
+pipeline {
+  agent { label 'linux' }
+
+  options {
+    disableConcurrentBuilds()
+    /* Necessary for logos-side-builder local_folder source type. */
+    checkoutToSubdirectory('src/')
```
Comment (jakubgs) [inline Jenkinsfile:7]: You don't need the trailing slashes.

**Review by jakubgs — APPROVED** [review]
LGTM, but remember to add context to the commit body.

**Comment by serhanwbahar** [comment]
> LGTM, but remember to add context to the commit body.

done.

**Pull request merged**